### PR TITLE
scripts: introduce configurable number of load gen threads

### PR DESCRIPTION
**Pull request by t-lo**
This change adds support for a --threads flag to `wrk2/render.sh`, and
adds the number of benchmark load generator threads to `benchmark.sh`
and `benchmark-multi.sh`.

Signed-off-by: Thilo Fromm <thilo@kinvolk.io>

### CODE REVIEW

File: scripts/istio/benchmark.sh [hunk 1]
```diff
@@ -21,7 +21,7 @@ istioctl=$(grok_cmd 5 "istioctl" $@)
 [ -z $istioctl ] && { echo "Aborting."; exit 1; }
 
 istio_type="tuned"
-[ $# -ge 6 ] && istio_type="$5"
+[ $# -ge 6 ] && istio_type="$6"
```
Comment (invidian) [inline scripts/istio/benchmark.sh:24]: This commit should be squashed with previous one.

File: README.md [hunk 2]
```diff
@@ -46,16 +46,17 @@ Before continuing, make sure the `KUBECONFIG` environment variable points to
...
 You're now set up to start a benchmark. You may specify the number of apps in
-the cluster, the benchmark run time, and the requests per second via the
+the cluster, the benchmark run time, the requests per second, and the number of
+parallel connections / threads used by the benchmark load generator via the
 command line.
 
 The following example starts a linkerd benchmark running 5 minutes, with 10
-apps and a constant request rate of 100RPS:
-* `scripts/linkerd/benchmark.sh 10 5m 100`
+apps a constant request rate of 100RPS, using 8 load generator threads:
```
Comment (invidian) [inline README.md:54]: I have a feeling, that number of positional parameters passed to benchmark scripts starts getting ridiculously long :) We should switch to named parameters at some point.

File: scripts/istio/benchmark-multi.sh [hunk 3]
```diff
@@ -17,25 +17,28 @@ KUBECONFIG=$(print_kubeconfig_path "$asset_dir")
     exit 1; }
 export KUBECONFIG
 
-nr_apps="10"
+nr_apps="4"
```
Comment (invidian) [inline scripts/istio/benchmark-multi.sh]: Changing defaults should go to another commit and be justified IMO

File: scripts/istio/benchmark.sh [hunk 4]
```diff
@@ -5,20 +5,23 @@ set -euo pipefail
 script_dir=$(dirname ${BASH_SOURCE[0]})
 source "$script_dir/../common.sh"
 
-nr_apps="10"
```
Comment (invidian) [inline scripts/istio/benchmark.sh]: Ditto here regarding defaults.

File: scripts/linkerd/benchmark-multi.sh [hunk 5]
```diff
@@ -17,25 +17,28 @@ KUBECONFIG=$(print_kubeconfig_path "$asset_dir")
     exit 1; }
 export KUBECONFIG
 
-nr_apps="10"
```
Comment (invidian) [inline scripts/linkerd/benchmark-multi.sh]: Ditto here regarding defaults.

File: scripts/linkerd/benchmark.sh [hunk 6]
```diff
@@ -5,16 +5,19 @@ set -euo pipefail
 script_dir=$(dirname ${BASH_SOURCE[0]})
 source "$script_dir/../common.sh"
 
-nr_apps="10"
```
Comment (invidian) [inline scripts/linkerd/benchmark.sh]: Ditto here regarding defaults.

Comment (invidian) [reply scripts/linkerd/benchmark.sh]: Actually, what about having those defaults in single file which can be sourced?

**Review by invidian — APPROVED** [review]
I didn't run it, but seems OK. I left some comments with potential improvements.

File: scripts/istio/benchmark.sh (hunk 1 above)
Comment (t-lo) [reply scripts/istio/benchmark.sh:24]: There's pros and cons to that, I personally prefer reading the full history of a PR.

File: README.md (hunk 2 above)
Comment (t-lo) [reply README.md:54]: Oh yes, and the potential squash target you raised above is a good illustration for this. We should probably switch to (getopts)[https://sookocheff.com/post/bash/parsing-bash-script-arguments-with-shopts/] at some point.

File: scripts/istio/benchmark-multi.sh (hunk 3 above)
Comment (t-lo) [reply scripts/istio/benchmark-multi.sh]: Like,

> The change furthermore introduces sane defualts for RPS, app count,
and nr of threads when run on a minimal cluster of 2 t1.small workers.

?

I agree it should go into a separate commit though.

File: scripts/linkerd/benchmark.sh (hunk 6 above)
Comment (t-lo) [reply scripts/linkerd/benchmark.sh]: Yupp, good idea. Patches welcome :)

File: scripts/istio/benchmark-multi.sh (hunk 3 above)
Comment (t-lo) [reply scripts/istio/benchmark-multi.sh]: I've removed the updated defaults from this PR and will issue a separate PR for that. Thank you for calling this out!.

File: scripts/istio/benchmark.sh (hunk 4 above)
Comment (t-lo) [reply scripts/istio/benchmark.sh]: I've removed the updated defaults from this PR and will issue a separate PR for that. Thank you for calling this out!.

File: scripts/linkerd/benchmark-multi.sh (hunk 5 above)
Comment (t-lo) [reply scripts/linkerd/benchmark-multi.sh]: I've removed the updated defaults from this PR and will issue a separate PR for that. Thank you for calling this out!.

File: scripts/linkerd/benchmark.sh (hunk 6 above)
Comment (t-lo) [reply scripts/linkerd/benchmark.sh]: I've removed the updated defaults from this PR and will issue a separate PR for that. Thank you for calling this out!.

File: scripts/istio/benchmark.sh (hunk 1 above)
Comment (invidian) [reply scripts/istio/benchmark.sh:24]: Well, for me the strong argument for it is that it breaks "every commit builds" rule, which in some cases (bisecting?) is a nice to have, even if it's not enforced by the project. But ¯\\_(ツ)_/¯

File: README.md (hunk 2 above)
Comment (invidian) [reply README.md:54]: Or https://stackoverflow.com/a/14203146 :)

File: scripts/istio/benchmark-multi.sh (hunk 3 above)
Comment (invidian) [reply scripts/istio/benchmark-multi.sh]: Hm, changes are still there. Are you decided to add them to the first commit then?

**Review by invidian — COMMENTED** [review]
I don't like changing values back end forth in a scope of single PR. I'd squash all commits and update it's message, so we have clear history of master branch.

File: scripts/istio/benchmark-multi.sh (hunk 3 above)
Comment (t-lo) [reply scripts/istio/benchmark-multi.sh]: This change was removed with the last commit - I'll squash the commits in this PR to ease review.

File: scripts/common.sh [hunk 7]
```diff
@@ -290,8 +290,9 @@ function run_benchmark() {
     local instances=$2
     local inject=$3
     local log_dir="${BENCHMARK_LOGFILE_DIR:-.}/"
-    local duration=${4:-30m}
-    local rate=${5:-800}
+    local duration=${4:-10m}
```
Comment (invidian) [inline scripts/common.sh]: 30m -> 10m and 800 -> 100, do we still want to change that? Maybe it should be mentioned in commit message or go into separate commit?

**Review by invidian — COMMENTED** [review]
Just one last commit. Other than that LGTM, though I didn't test.

File: scripts/common.sh (hunk 7 above)
Comment (t-lo) [reply scripts/common.sh]: Fixed, good catch!

**Review by invidian — APPROVED** [review]
LGTM, though I didn't test it. I'm going to do it now, but it should be good to merge.
